src/base/algorithm.h

```diff
@@ -28,14 +28,42 @@
 
 #pragma once
 
-namespace Dict
+#include <type_traits>
+
+namespace Algorithm
 {
-    // To be used with QMap, QHash and it's variants
-    template <typename Dictionary, typename BinaryPredicate>
-    void removeIf(Dictionary &&dict, BinaryPredicate p)
+    template <typename ...>
+    using void_t = void;  // replace this with std::void_t in C++17
+
+    template <typename T, typename = void>
+    struct HasMappedType
+        : std::false_type
+    {
+    };
+
+    template <typename T>
+    struct HasMappedType<T, void_t<typename T::mapped_type>>
+        : std::true_type
+    {
+    };
+
+    // To be used with usual iteratable types, such as QVector, QSet, etc.
+    template <typename T, typename BinaryPredicate
+        , typename std::enable_if_t<!HasMappedType<T>::value, int> = 0>
```

Comment:
Shouldn't it be
```c++
typename std::enable_if_t<!HasMappedType<T>::value, int>* = nullptr
```
Note the asterisk after type name!

Although I would prefer a different syntax:
```с++
typename = std::enable_if_t<!HasMappedType<T>::value, int>
```

Reply:
>Shouldn't it be
typename std::enable_if_t<!HasMappedType<T>::value, int>* = nullptr

It can, but why? what is the advantage?

>Although I would prefer a different syntax:
typename = std::enable_if_t<!HasMappedType<T>::value, int>

From what I read on https://en.cppreference.com/w/cpp/types/enable_if that won't work, I'm referring to the example in notes section.

Reply:
>It can, but why? what is the advantage?

Oops... I didn't notice you explicitly set the second parameter of enable_if to int. Then it's correct.
>that won't work, I'm referring to the example in notes section.

You're right.